File: Simulacao/MM_2/atores/cliente.js
class Clientes {

    constructor() {
        this.clientes = [];
        this.numCliente = 0;
    }

    addCliente(TR) {
        this.numCliente += 1;
        this.clientes.push(new Cliente(this.numCliente, TR));
    }

    minHs() {
        var minHs = 999999;
        
        var HSvalues = this.clientes.filter(cliente =>
            cliente.HS != null && cliente.HS != undefined);
            
        if (HSvalues && HSvalues.length > 0){
            minHs = HSvalues.reduce((min, cliente) =>
                Math.min(min, cliente.HS), HSvalues[0].HS);
        }

        return minHs;
    }

    removeCliente() {
        var minHS = this.minHs();
        
        const index = this.clientes.findIndex(cliente => cliente.HS != null && cliente.HS != undefined && cliente.HS == minHS);

        const clienteAux = this.clientes[index];
        this.clientes.splice(index, 1);

        return clienteAux;
    }

    setNextCliente(HS, TS, index) {
        var cliente = this.clientes.find(cliente => !cliente.HS);
        cliente.HS = HS;
        cliente.TS = TS;
        cliente.indexOperador = index;
    }

}

class Cliente {

    constructor(numCliente, TR){
        this.numCliente = numCliente;
        this.TR = TR;
        this.TS = null;
        this.HS = null;
        this.indexOperador = null;
    }

    getInicioServico() {
        return this.HS - this.TS;
    }

    getTempoFila() {
        return this.getInicioServico() - this.TR;
    }

}

File: Simulacao/MM2_Python/Cliente.py
class Cliente:
    def __init__(self):
        self.clientes = []
        self.numCliente = 0

    def addCliente(self, TR):
        self.numCliente = self.numCliente + 1
        self.clientes.append(clientes(self.numCliente, TR))

    def lessHS(self):
        aux = 999999
        for x in range(len(self.clientes)):
            if self.clientes[x].HS != None:
                if self.clientes[x].HS <= aux:
                    aux = self.clientes[x].HS
        return aux

    def removeCliente(self):
        aux = 999999
        for x in range(len(self.clientes)):
            if self.clientes[x].HS != None:
                if self.clientes[x].HS <= aux:
                    aux = self.clientes[x].HS
                    index = x
        return self.clientes.pop(index)

    def setNextClient(self, HS, TS, index):
        for x in self.clientes:
            if x.HS == None:
                x.setHS(HS)
                x.setTS(TS)
                x.setOperador(index)
                return

class clientes:
    def __init__(self, numCliente, TR):
        self.numCliente = numCliente
        self.TR = TR
        self.TS = None
        self.HS = None
        self.indexOperador = None

    def setOperador(self, index):
        self.indexOperador = index
    def setTS(self, TS):
        self.TS = TS
    def setHS(self, HS):
        self.HS = HS

    def getInicioServico(self):
        return self.HS - self.TS
    def getTempoFila(self):
        return self.getInicioServico() - self.TR